x-pack/plugins/ml/public/url_generator.ts

```diff
@@ -64,39 +208,265 @@ export class MlUrlGenerator implements UrlGeneratorsDefinition<typeof ML_APP_URL
 
   public readonly id = ML_APP_URL_GENERATOR;
 
-  public readonly createUrl = async ({ page, ...params }: MlUrlGeneratorState): Promise<string> => {
-    if (page === 'explorer') {
-      return this.createExplorerUrl(params);
+  public readonly createUrl = async (mlUrlGeneratorState: MlUrlGeneratorState): Promise<string> => {
+    switch (mlUrlGeneratorState.page) {
+      case ML_PAGES.ANOMALY_DETECTION:
+        return this.createAnomalyDetectionJobManagementUrl(
+          mlUrlGeneratorState as AnomalyDetectionUrlState
+        );
+      case ML_PAGES.ANOMALY_EXPLORER:
+        return this.createExplorerUrl(mlUrlGeneratorState as ExplorerUrlState);
+      case ML_PAGES.TIME_SERIES_EXPLORER:
+        return this.createSingleMetricViewerUrl(mlUrlGeneratorState as TimeSeriesExplorerUrlState);
+      case ML_PAGES.DATA_FRAME_ANALYTICS:
+        return this.createDataframeAnalyticsUrl(mlUrlGeneratorState as DataFrameAnalyticsUrlState);
+      case ML_PAGES.DATA_FRAME_ANALYTICS_EXPLORATION:
+        return this.createDataframeAnalyticsExplorationUrl(
+          mlUrlGeneratorState as DataFrameAnalyticsExplorationUrlState
+        );
+      case ML_PAGES.DATA_VISUALIZER:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_FILE:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_INDEX_SELECT:
+        return this.createDataVisualizerUrl(mlUrlGeneratorState as DataVisualizerUrlState);
+      case ML_PAGES.DATA_VISUALIZER_NEW_JOB:
+        return this.createIndexBasedMlUrl(mlUrlGeneratorState as MlGenericUrlState);
+      case ML_PAGES.DATA_VISUALIZER_VIEWER:
+        return this.createIndexBasedMlUrl(mlUrlGeneratorState as MlGenericUrlState);
```

Comment:
please remove manual typecasting, it's redundant

Reply:
I think typescript is failing on this one. I type-casted here because TS wasn't recognizing it properly.

Reply:
ha, that's an issue with IntelliJ, for some reason it wasn't highlighting an error and showing my correct type-casting 🤔  probably TS server ran out of memory